HLTrigger/Muon/interface/HLTTriMuonIsolation.h

```diff
@@ -0,0 +1,242 @@
+{
+    //register products
+    produces<reco::CompositeCandidateCollection>("Taus");
+    produces<reco::CompositeCandidateCollection>("SelectedTaus");
+}
+
+HLTTriMuonIsolation::~HLTTriMuonIsolation(){ }
+
+void
+HLTTriMuonIsolation::produce(edm::StreamID sid, edm::Event & iEvent, edm::EventSetup const & iSetup) const
+{
+    std::unique_ptr<reco::CompositeCandidateCollection> Taus        (new reco::CompositeCandidateCollection);
+    std::unique_ptr<reco::CompositeCandidateCollection> SelectedTaus(new reco::CompositeCandidateCollection);
+
+    // Get the L3 muon candidates
+    edm::Handle<reco::RecoChargedCandidateCollection> L3MuCands;
+    iEvent.getByToken(L3MuonsToken_, L3MuCands);
+
+    // Get the L3 muon candidates that passed the filter
+    edm::Handle<trigger::TriggerFilterObjectWithRefs> L3DiMuonsFilterCands;
+    iEvent.getByToken(L3DiMuonsFilterToken_, L3DiMuonsFilterCands);
+    
+    std::vector<reco::RecoChargedCandidateRef> PassedL3Muons;
+    L3DiMuonsFilterCands->getObjects(trigger::TriggerMuon, PassedL3Muons);
+  
+    // Get the Trk + L3 muon candidates (after merging)
+    edm::Handle<reco::RecoChargedCandidateCollection> AllMuCands;
+    iEvent.getByToken(AllMuonsToken_, AllMuCands);
+    
+    // Get iso tracks
+    edm::Handle<reco::TrackCollection> IsoTracks;
+    iEvent.getByToken(IsoTracksToken_, IsoTracks);
+          
+    if (AllMuCands->size() >= 3 && L3MuCands->size() >= 2){
+        // Create the 3-muon candidates
+        // loop over L3/Trk muons and create all combinations
+        auto AllMuCands_end = AllMuCands->end();
+        for (auto i = AllMuCands->begin(); i != AllMuCands_end; ++i) {
+            // check that muon_i passes the previous filter
+            bool passingPreviousFilter_1 = false;
+            const reco::TrackRef &tk_i = i->track();
+            for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
+                reco::TrackRef candTrkRef = (*imu)->get<reco::TrackRef>();
+                if (reco::deltaR2(tk_i->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_1 = true;
+            }
+            for (auto j = i+1; j != AllMuCands_end; ++j) {
+                // check that muon_j passes the previous filter
+                bool passingPreviousFilter_2 = false;
+                const reco::TrackRef &tk_j = j->track();
+                for (std::vector<reco::RecoChargedCandidateRef>::const_iterator imu = PassedL3Muons.begin(); imu != PassedL3Muons.end(); ++imu){
+                    reco::TrackRef candTrkRef = (*imu)->get<reco::TrackRef>();
+                    if (reco::deltaR2(tk_j->momentum(), candTrkRef->momentum()) < (0.03*0.03)) passingPreviousFilter_2 = true;                
+                }
+                for (auto k = j+1; k != AllMuCands_end; ++k){
```

Comment:
No need to execute 3rd loop if passingPreviousFilter_1 and passingPreviousFilter_2 are both FALSE (or 0).